### PR TITLE
feat(tts): add MeCab dictionary support and enhance multilingual text…

### PR DESCRIPTION
**Pull request by freddy311082**
## 🎯 What problem does this PR solve?

- Chatterbox multilingual hallucinated audio for Japanese (and Chinese) because the tokenizer mapped CJK kanji/hanzi to `[UNK]`.
- Existing character-level kanji preprocessing was not enough: real-world Japanese needs word-level morphological analysis to produce correct phonetic readings.
- Text preprocessing logic in C++ was a loose collection of free functions and a raw `mecab_t*`-style approach was being prototyped, which is hard to maintain and leaks resources.

## 📝 How does it solve it?

- Integrate **MeCab + IPAdic** for Japanese: at runtime the addon morphologically segments the input, extracts each token's reading from MeCab's feature string, and converts katakana to hiragana before tokenization. Result: kanji never reach the tokenizer as `[UNK]`.
- Wire MeCab cross-platform via **vcpkg** (`mecab` added to `vcpkg.json` and `vcpkg-configuration.json`, linked in both desktop and mobile CMake targets and in the unit test target).
- Ship the **precompiled IPAdic dictionary inside the package**: `dict/mecab-ipadic/{char.bin, dicrc, matrix.bin, mecabrc, sys.dic, unk.dic}` (~50 MB). A minimal `mecabrc` is included so MeCab can initialize without a system-wide rc. Endianness is the same on all our target platforms (Android, iOS, Windows, Linux x86/x64, macOS), so a single compiled dictionary works everywhere.
- Add `npm run build:mecab-dict` (`scripts/build-mecab-dict.sh`) to recompile the dictionary from IPAdic source using `mecab-dict-index` (only needed when refreshing the dictionary).
- Plumb the dictionary location end-to-end:
  - JS resolves a fixed location relative to the package: `path.join(__dirname, 'dict')`.
  - New `dictPath` field on `ChatterboxConfig`, propagated through `AddonJs`, `TTSModel` and `ChatterboxEngine::loadTextPreprocessor(tokenizerPath, dictPath)`.
  - The Cangjie table for `zh` keeps being resolved next to the tokenizer.
- Refactor text preprocessing for clarity and RAII:
  - All free functions (`decomposeKoreanToJamo`, `convertKatakanaToHiragana`, `convertChineseToCangjie`, `convertJapaneseWithMeCab`, `loadCangjieTable`, `loadMeCab`, `preprocess`) moved as methods of a new `class ChatterboxTextPreprocessor`.
  - The MeCab handle is now a private `std::unique_ptr<mecab_t, MeCabDeleter>`, replacing the raw pointer / handle struct.
  - Long loops split into focused helpers (`detectSequenceLength`, `extractLeadingBits`, `decodeCodepointAt` for UTF-8 decoding; `isContentNode`, `hasReading`, `appendNodeReading`, `buildHiraganaFromNodes` for MeCab traversal).
  - `decodeUtf8` and `encodeCodepoint` are now `static` utility methods.
  - `ChatterboxEngine` holds a single `ChatterboxTextPreprocessor` instead of separate Cangjie / MeCab members.

## 🧪 How was it tested?

- C++ unit tests: **197 pass / 9 skipped (require optional env or LavaSR models) / 0 fail** (`./build/addon/test/unit/qvac-lib-inference-tts-unit-test`). Includes 23 preprocessing-specific tests (Utf8Test, KoreanJamoTest, KatakanaHiraganaTest, ChineseCangjieTest, PreprocessDispatchTest).
- JS unit tests: **34/34 pass, 102/102 asserts** (`npm run test:unit`).
- Lint: `npm run lint` passes.
- Integration test added: `test/integration/test-zh-ja.js`. Loads the Chatterbox multilingual model with `language: 'ja'`, synthesizes `こんにちは世界。今日はいい天気です。`, validates audio sample/duration ranges, and writes the WAV for manual verification. Run successful end-to-end with the bundled MeCab dictionary.
- Manual transcription of the generated WAV (Whisper / Google Translate) confirms the audio matches the input text, fixing the previous hallucination.

## 🔌 API Changes

- New `dictPath` field on the native `ChatterboxConfig`, passed automatically by the JS layer; consumers using the high-level `ONNXTTS` JS API don't need to pass anything (the path is resolved relative to the package).

```typescript
// Native ChatterboxConfig (C++) now exposes:
struct ChatterboxConfig {
  std::string language;
  std::vector<float> referenceAudio;
  std::string tokenizerPath;
  std::string speechEncoderPath;
  std::string embedTokensPath;
  std::string conditionalDecoderPath;
  std::string languageModelPath;
  std::string dictPath; // new: location of the runtime dictionaries (e.g. mecab-ipadic)
  bool lazySessionLoading = false;
  bool useGPU = false;
};

### CODE REVIEW

File: packages/qvac-lib-infer-onnx-tts/package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@qvac/tts-onnx",
-  "version": "0.8.3",
+  "version": "0.9.0",
```
Comment (GustavoA1604) [inline packages/qvac-lib-infer-onnx-tts/package.json]: You can bump to 0.8.4 only as this is not a breaking change

File: packages/tts-onnx/package.json [hunk 2]
```diff
@@ -37,6 +38,7 @@
     "binding.js",
     "index.js",
     "lib",
+    "dict",
```
Comment (GustavoA1604) [inline packages/tts-onnx/package.json]: Could you keep the dict downloadable instead of put in the package? It will make the package several MB bigger but will only help for japanese TTS

We can add code in ìndex.js` saying that, if japanese language is selected, then we need to pass dictPath parameter. Also @ishanvohra2 can help to add necessary code to SDK/models to registry to facilitate download

File: packages/qvac-lib-infer-onnx-tts/index.js [hunk 3]
```diff
@@ -456,6 +446,24 @@ class ONNXTTS {
...
   }
 
+  _getChatterboxTtsParams () {
+    const params = {
+      tokenizerPath: this._tokenizerPath || '',
+      speechEncoderPath: this._speechEncoderPath || '',
+      embedTokensPath: this._embedTokensPath || '',
+      conditionalDecoderPath: this._conditionalDecoderPath || '',
+      languageModelPath: this._languageModelPath || '',
+      dictPath: CHATTERBOX_DICT_DIR,
```
Comment (GustavoA1604) [inline packages/qvac-lib-infer-onnx-tts/index.js]: Shouldn't we name this something like mecabDictPath? Since its not any dict

Comment (GustavoA1604) [reply packages/qvac-lib-infer-onnx-tts/index.js]: Also, remember to update index.d.ts with the API changes

File: packages/qvac-lib-infer-onnx-tts/test/integration/test-zh-ja.js
Comment (GustavoA1604) [inline packages/qvac-lib-infer-onnx-tts/test/integration/test-zh-ja.js:1]: In TTS we are keeping a single addon.test.js file, could you add the test there? Instead of creating a new file

File: packages/qvac-lib-infer-onnx-tts/test/integration/test-zh-ja.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  return path.join(modelDir, `${baseName}.onnx`)
+}
+
+const TEXT_JA = 'こんにちは世界。今日はいい天気です。'
+
+const EXPECTATION = {
+  minSamples: 5000,
+  maxSamples: 5000000,
+  minDurationMs: 200,
+  maxDurationMs: 300000
```
Comment (GustavoA1604) [inline packages/qvac-lib-infer-onnx-tts/test/integration/test-zh-ja.js]: Shouldn't we have more strict settings? Considering that before audio was also generated, but was broken

File: packages/qvac-lib-infer-onnx-tts/addon/src/model-interface/ChatterboxTextPreprocessor.cpp [hunk 5]
```diff
@@ -185,24 +250,93 @@ CangjieTable loadCangjieTable(const std::string &tsvPath) {
...
-      table[charCp[0]] = code;
+    if (charCp.size() == 1 &&
+        cangjieTable_.find(charCp[0]) == cangjieTable_.end()) {
+      cangjieTable_[charCp[0]] = code;
     }
   }
+}
+
+void ChatterboxTextPreprocessor::loadMeCab(const std::string& dicPath) {
+  std::string rcPath = dicPath + "/mecabrc";
```
Comment (GustavoA1604) [inline packages/qvac-lib-infer-onnx-tts/addon/src/model-interface/ChatterboxTextPreprocessor.cpp]: Need to handle dicPath using appropriate structs instead of plain string. Else it will fail on windows if the path contains a space, for instance

File: packages/tts-onnx/addon/src/model-interface/ChatterboxTextPreprocessor.cpp [hunk 6]
```diff
@@ -185,24 +250,93 @@ CangjieTable loadCangjieTable(const std::string &tsvPath) {
...
+
+} // namespace
 
-  return table;
+void ChatterboxTextPreprocessor::appendNodeReading(
+    const mecab_node_t* node, std::string& result) const {
+  std::string reading = extractReading(node->feature);
+  if (hasReading(reading)) {
+    result += convertKatakanaToHiragana(reading);
+    return;
```
Comment (GustavoA1604) [inline packages/tts-onnx/addon/src/model-interface/ChatterboxTextPreprocessor.cpp:333]: Could we assert that there is a reading for japanese? Else, if dict is corrupt somehow, it will simply go on with the flow with no error messages, which we know will break

**Review by GustavoA1604 — CHANGES_REQUESTED** [review]
Please ask cursor to update NOTICE file considering new library dependency

File: packages/qvac-lib-infer-onnx-tts/addon/test/unit/src/ChatterboxTextPreprocessorTest.cpp [hunk 7]
```diff
@@ -75,122 +80,102 @@ TEST_F(KoreanJamoTest, handlesMixedText) {
...
-  CangjieTable table_;
-
-  void SetUp() override {
-    table_[0x65E5] = "a";
-    table_[0x6708] = "b";
-  }
+  Preprocessor preprocessor_;
 };
 
 TEST_F(ChineseCangjieTest, convertsCjkCharacter) {
```
Comment (ogad-tether) [inline packages/qvac-lib-infer-onnx-tts/addon/test/unit/src/ChatterboxTextPreprocessorTest.cpp:135]: This test now asserts hello to hello and does not validate CJK to Cangjie behavior. Could we restore a direct assertion that at least one Chinese codepoint is mapped after loading a table, and add zh dispatch coverage to avoid regressions?

File: packages/tts-onnx/vcpkg.json [hunk 8]
```diff
@@ -16,6 +16,7 @@
       "name": "tokenizers-cpp",
       "version>=": "0.1.1#2"
     },
+    "mecab",
```
Comment (ogad-tether) [inline packages/tts-onnx/vcpkg.json:17]: Since this adds mecab and the package now ships a bundled IPAdic dictionary, can we also update NOTICE and related attribution metadata for these third-party assets before release?

**Comment by ogad-tether** [comment]
Review summary: (1) Chinese preprocessing tests were weakened and no longer verify Cangjie mapping behavior, and (2) MeCab/IPAdic attribution appears missing from NOTICE despite adding and shipping these assets.

**Comment by freddy311082** [comment]
CI update result que be seen here: https://github.com/tetherto/qvac/actions/runs/25825140756

**Review by ishanvohra2 — DISMISSED** [review]

**Review by GustavoA1604 — CHANGES_REQUESTED** [review]
Please help fix what was suggested by the comment from Omar, plus

1. Bump to 0.10.0 instead to avoid surprises in existing published SDK 0.11.0

2. Debug std::cerr statements left in production code

ChatterboxEngine.cpp and ChatterboxTextPreprocessor.cpp are littered with debug prints that write to stderr on every synthesis call:

std::cerr << ">>> [TOKENIZE-CPP] raw input='" << text << "' lang=" << language_ ...
std::cerr << ">>> [MECAB-CPP] convertJapaneseWithMeCab input='" << text << "' ...
std::cerr << ">>> [ENGINE-CPP] Loading MeCab dictionary from: " ...
There are ~10+ of these. They will flood stderr in production for all users regardless of language. They should be replaced with QLOG calls (already used elsewhere in the same file) or removed.

3. prepareTextForTokenization is now a behavioral breaking change for all non-English languages

The old behavior simply prepended [lang]:

// OLD
return "[" + language + "]" + text;
The new behavior lowercases (NFKD) and replaces spaces with [SPACE] for all non-English languages (Spanish, Portuguese, French, German, etc.):

// NEW
const std::string normalized = applyLowercaseNfkd(text);
const std::string prefixed = "[" + language + "]" + normalized;
return replaceSpacesWithToken(prefixed);
This silently changes synthesis behavior for every existing user of non-English Chatterbox. For example, "Hola Mundo" in Spanish now becomes "[es]hola[SPACE]mundo" instead of "[es]Hola Mundo". The test that previously asserted "[es]Hola mundo" was updated to expect the new output, masking the regression rather than raising it. This change may or may not be correct for the model, but it needs explicit justification and shouldn't be bundled silently.

4. maxSpeechTokens hardcoded to MAX_NEW_TOKENS_SPEECH (1024) for all languages

// OLD — proportional to input length
int maxSpeechTokens = std::max(MIN_SPEECH_TOKENS, textTokenCount * SPEECH_TO_TEXT_MAX_RATIO);
maxSpeechTokens = std::min(maxSpeechTokens, MAX_NEW_TOKENS_SPEECH);
// NEW — always 1024, for every language
int maxSpeechTokens = MAX_NEW_TOKENS_SPEECH;
For short inputs (e.g. a 5-word sentence ~10 tokens), the old cap would have been max(50, 30) = 50. Now it's always 1024. This changes generation behavior and potentially performance for all languages, not just Japanese. This looks like a workaround for hallucination in Japanese that has a side-effect on every language.

5. /FORCE:MULTIPLE linker flag on Windows

target_link_options(${inference-tts}_module PRIVATE /FORCE:MULTIPLE)
This suppresses multiply-defined symbol linker errors rather than resolving the actual conflict. This is a known code smell — it can silently pick one symbol implementation arbitrarily, leading to subtle undefined behavior. The underlying symbol conflict with MeCab should be identified and properly resolved (e.g., with OBJECT library separation or namespace isolation).

6. Default mecabDictPath fallback may not be implemented

The index.d.ts, README, and CHANGELOG all say the default is <package>/dict when mecabDictPath is omitted. But index.js just passes this._mecabDictPath || '' to native code. An empty string results in std::filesystem::path("") / "mecab-ipadic" = "mecab-ipadic" (a relative path), which will fail at runtime for most users. The auto-resolution to path.join(__dirname, 'dict') that the PR description promises doesn't appear to be in the diff.

7. Chinese Cangjie unit tests removed entirely

Commit d46a21f drops the entire ChineseCangjieTest suite (5 tests). The zh preprocessing code path (convertChineseToCangjie) still exists and is reachable. This was flagged by @ogad-tether and remains unaddressed. Removing tests for live code creates a silent regression risk.

8. isStopTokenInTopK affects all language synthesis quality

The new forced-stop mechanism (STOP_FORCE_TOPK = 20) fires whenever the stop token is in the top-20 logits. While this may reduce hallucination in Japanese, it's applied unconditionally to all languages and could affect the naturalness of synthesis in languages that previously worked well. There's no ablation or justification for why 20 is the right threshold.

9. fmt and spdlog are added to vcpkg-configuration.json but don't appear to be used by any file in this diff — worth confirming they're needed.

10. The PR description still refers to dictPath in ChatterboxConfig (the struct example in the body), but the actual field name in code is mecabDictPath. Minor but could confuse reviewers.

**Review by GustavoA1604 — DISMISSED** [review]

**Review by ogad-tether — APPROVED** [review]

**Review by ogad-tether — DISMISSED** [review]

**Comment by freddy311082** [comment]
@GustavoA1604 CI passed here: https://github.com/tetherto/qvac/actions/runs/26251826129/job/77266193453
